artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/LastValueQueue.java

```diff
@@ -350,25 +342,33 @@ public SimpleString getLastValueKey() {
       return Collections.unmodifiableSet(map.keySet());
    }
 
+   @Override
+   public boolean canBeDelivered(MessageReference ref) {
+      if (isNonDestructive() && ref instanceof HolderReference) {
+         return !((HolderReference)ref).isDelivered();
+      } else {
+         return true;
+      }
+   }
+
    private static class HolderReference implements MessageReference {
 
       private final SimpleString prop;
 
-      private volatile boolean delivered = false;
+      private volatile AtomicBoolean delivered = new AtomicBoolean(false);
```

Comment:
Please use the AtomicLongFieldUpdater instead?

Reply:
As far as I can tell you can't use `AtomicLongFieldUpdater` on a `boolean`.